### PR TITLE
[SYCL] Define SYCL_DISABLE_FALLBACK_ASSERT macro

### PR DESCRIPTION
**Pull request by elizabethandrews**
Define SYCL_DISABLE_FALLBACK_ASSERT for FPGA

Signed-off-by: Elizabeth Andrews <elizabeth.andrews@intel.com>

### CODE REVIEW

File: clang/lib/Frontend/InitPreprocessor.cpp [hunk 1]
```diff
@@ -1198,8 +1198,10 @@ static void InitializePredefinedMacros(const TargetInfo &TI,
         DeviceSubArch != llvm::Triple::SPIRSubArch_fpga)
       Builder.defineMacro("SYCL_USE_NATIVE_FP_ATOMICS");
     // Enable generation of USM address spaces for FPGA.
-    if (DeviceSubArch == llvm::Triple::SPIRSubArch_fpga)
+    if (DeviceSubArch == llvm::Triple::SPIRSubArch_fpga) {
       Builder.defineMacro("__ENABLE_USM_ADDR_SPACE__");
+      Builder.defineMacro("SYCL_DISABLE_FALLBACK_ASSERT");
```
[the remark this answers is not in the document]
Comment (s-kanaev) [reply clang/lib/Frontend/InitPreprocessor.cpp:1203]: @bader , @elizabethandrews 
There's a plan to re-enable fallback assert by default.
Hence, I've re-opened PR.

[the remark this answers is not in the document]
Comment (dm-vodopyanov) [reply clang/lib/Frontend/InitPreprocessor.cpp:1203]: @s-kanaev , is some additional work expected in this PR? Can the PR be merged?

[the remark this answers is not in the document]
Comment (s-kanaev) [reply clang/lib/Frontend/InitPreprocessor.cpp:1203]: I'll check with @elizabethandrews  what sort of test should be added here.

**Review by smanna12 — CHANGES_REQUESTED** [review]
>>Can you add test?

Tests have been already added. Sorry I missed that.

**Review by smanna12 — APPROVED** [review]

**Comment by s-kanaev** [comment]
@premanandrao , a friendly ping

File: clang/lib/Frontend/InitPreprocessor.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (bader) [reply clang/lib/Frontend/InitPreprocessor.cpp:1203]: @s-kanaev, what is overall plan for fallback assert? Are we going to commit #4824 and close #4694?

[the remark this answers is not in the document]
Comment (s-kanaev) [reply clang/lib/Frontend/InitPreprocessor.cpp:1203]: Absolutely, #4694 is going to be closed. #4824 and #4780 are going to be committed. There were two types of issues:
 * enabling assertions for interop kernels
 * impact of assertions on FPGA
The former is fixed in #4712.
The latter is fixed with help of #4824 and #4780.

[the remark this answers is not in the document]
Comment (bader) [reply clang/lib/Frontend/InitPreprocessor.cpp:1203]: I think the second issue impacts non-FPGA devices as well. AFAIK, additional assert-specific device code is emitted for all devices even though assert is not used. Is it right? If so, what is the plan to address the overhead associated with the additional code?

[the remark this answers is not in the document]
Comment (s-kanaev) [reply clang/lib/Frontend/InitPreprocessor.cpp:1203]: > AFAIK, additional assert-specific device code is emitted for all devices even though assert is not used. Is it right?
Yes.
To overcome this impact I need to do some refactoring which is a bit stuck at limitations of sycl-post-link tool.
The limitation could be handled with employing of `__sycl_kernel` attribute, though, it requires for the function to be template which isn't acceptable at the moment. Also, @AlexeySachkov  can't approve the w/a in sycl-post-link I presented to him.
The overall idea of refactoring can be caught with changes of devicelib in #4987.

[the remark this answers is not in the document]
Comment (elizabethandrews) [reply clang/lib/Frontend/InitPreprocessor.cpp:1203]: > 

There is a test in this patch for the macro. From a FE perspective, I don't think further tests are required

**Comment by elizabethandrews** [comment]
> I'll check with @elizabethandrews what sort of test should be added here

There is a test in this patch for the macro. From a FE perspective, I don't think further tests are required

File: clang/lib/Frontend/InitPreprocessor.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (bader) [reply clang/lib/Frontend/InitPreprocessor.cpp:1203]: Please, follow https://chris.beams.io/posts/git-commit/ and add proper PR title and description to #4987.

**Review by bader — APPROVED** [review]

**Review by premanandrao — APPROVED** [review]
LGTM

**Pull request merged**